spark/v3.3/spark/src/test/java/org/apache/iceberg/spark/sql/TestFilterPushDown.java

```diff
@@ -0,0 +1,465 @@
+    sql("INSERT INTO %s VALUES (3, 300, null)", tableName);
+
+    withDefaultTimeZone(
+        "UTC",
+        () -> {
+          checkCompleteFilterPushdown(
+              "t IS NULL" /* query predicate */,
+              "t IS NULL" /* Iceberg scan filters */,
+              ImmutableList.of(row(3, 300, null)));
+
+          // strict/inclusive projections for t < TIMESTAMP '2022-01-01T00:00:00.000Z' are equal
+          // so this filter selects entire partitions and can be pushed down completely
+          checkCompleteFilterPushdown(
+              "t < TIMESTAMP '2022-01-01T00:00:00.000Z'" /* query predicate */,
+              "t IS NOT NULL, t < 1640995200000000" /* Iceberg scan filters */,
+              ImmutableList.of(
+                  row(1, 100, Timestamp.from(Instant.parse("2021-06-30T01:00:00.000Z"))),
+                  row(2, 200, Timestamp.from(Instant.parse("2021-06-30T02:00:00.000Z")))));
+
+          // strict/inclusive projections for t < TIMESTAMP '2021-06-30T03:00:00.000Z' are NOT equal
+          // so this filter does NOT select entire partitions and can't be pushed down completely
+          checkFilterPushdown(
+              "t < TIMESTAMP '2021-06-30T03:00:00.000Z'" /* query predicate */,
+              "t < 2021-06-30 03:00:00" /* post scan filter */,
+              "t IS NOT NULL, t < 1625022000000000" /* Iceberg scan filters */,
+              ImmutableList.of(
+                  row(1, 100, Timestamp.from(Instant.parse("2021-06-30T01:00:00.000Z"))),
+                  row(2, 200, Timestamp.from(Instant.parse("2021-06-30T02:00:00.000Z")))));
+        });
+  }
+
+  @Test
+  public void testFilterPushdownWithBucketTransform() {
+    sql(
+        "CREATE TABLE %s (id INT, salary INT, dep STRING)"
+            + "USING iceberg "
+            + "PARTITIONED BY (dep, bucket(8, id))",
+        tableName);
+
+    sql("INSERT INTO %s VALUES (1, 100, 'd1')", tableName);
+    sql("INSERT INTO %s VALUES (2, 200, 'd2')", tableName);
+
+    checkFilterPushdown(
+        "dep = 'd1' AND id = 1" /* query predicate */,
+        "id = 1" /* post scan filter */,
+        "dep IS NOT NULL, id IS NOT NULL, dep = 'd1'" /* Iceberg scan filters */,
+        ImmutableList.of(row(1, 100, "d1")));
+  }
+
+  @Test
+  public void testFilterPushdownWithTruncateTransform() {
+    sql(
+        "CREATE TABLE %s (id INT, salary INT, dep STRING)"
+            + "USING iceberg "
+            + "PARTITIONED BY (truncate(1, dep))",
+        tableName);
+
+    sql("INSERT INTO %s VALUES (1, 100, 'd1')", tableName);
+    sql("INSERT INTO %s VALUES (2, 200, 'd2')", tableName);
+
+    checkFilterPushdown(
+        "dep = 'd1'" /* query predicate */,
+        "dep = d1" /* post scan filter */,
+        "dep IS NOT NULL" /* Iceberg scan filters */,
+        ImmutableList.of(row(1, 100, "d1")));
+  }
+
+  @Test
+  public void testFilterPushdownWithSpecEvolution() {
```

Comment:
Another suggestion here could be to test Truncate (1) evolving to truncate (2) and making sure startsWith(char) can be fully pushed down after evolution.

Or a date version of Month => Day when the timestamp boundaries are still on the month lines

Reply:
Added both scenarios as separate tests.